compiler/rustc_smir/src/rustc_smir/context.rs

```diff
@@ -240,6 +255,18 @@ impl<'tcx> Context for TablesWrapper<'tcx> {
         tables.tcx.type_of(item.internal(&mut *tables)).instantiate_identity().stable(&mut *tables)
     }
 
+    fn def_ty_with_args(
+        &self,
+        item: stable_mir::DefId,
+        args: &GenericArgs,
+    ) -> Result<stable_mir::ty::Ty, Error> {
+        let mut tables = self.0.borrow_mut();
+        let args = args.internal(&mut *tables);
+        let def_ty = tables.tcx.type_of(item.internal(&mut *tables));
+        // FIXME(celinval): use try_fold instead to avoid crashing.
```

Comment:
How does this crash occur, any way to trigger it, if so can we solve it now ?

Reply:
I did some investigation. I couldn't find any easy fix, such as an internal API that we could use here. I've tried using `try_instantiate_and_normalize_erasing_regions`, but the first thing it does is to invoke `instantiate()` so the same few crashes still happen. It's looking like we would either have to implement one ourselves, e.g., implement a `FallibleTypeFolder`, or extend the `EarlyBinder` to expose a `try_instantiate` method.

That said, the only two scenarios I manage to trigger a crash were:
1. The arguments length is less than the number of generic arguments.
2. There's an unexpected `GenericArgumentKind`.

From an API perspective, I would like to avoid crashes whenever possible, and that's why I wanted to make this API fallible. That said, it's looking pretty hard to do proper validation. I do wonder if we should remove the `Result<>` from this API. I think many of our APIs will unfortunately depend on the user doing the correct thing since at the MIR level we are working on, we expect types to have been checked already.

Reply:
Yeah your points are valid, this isn't easy to re-produce my main concern with this was if it was easy to re-produce or easy to crash. People could easy crash the compiler if they play with it long enough this shouldn't block this PR from passing.